Libraries/LibWebView/SourceHighlighter.cpp

```diff
@@ -18,11 +18,48 @@
 namespace WebView {
 
 SourceDocument::SourceDocument(StringView source)
-    : m_source(source)
 {
-    m_source.for_each_split_view('\n', AK::SplitBehavior::KeepEmpty, [&](auto line) {
-        m_lines.append(Syntax::TextDocumentLine { *this, line });
-    });
+    // HTML, CSS and JS differ slightly on what they consider a newline to be.
+    // In order to make them get along in documents that include a mix of the three, process the source to make the
+    // newlines consistent before doing any highlighting.
+    // We also replace any invalid characters.
+    // NOTE: We assume the source is utf-8 already, is this OK?
+
+    Utf8View const source_utf8 { source };
+
+    // Optimization: If all the newlines are \n, just use the input string.
+    if (!source_utf8.contains_any_of(Array<u32, 3> { '\r', 0x2028, 0x2029 })) {
+        m_source = String::from_utf8_with_replacement_character(source);
+    } else {
+        StringBuilder builder { source_utf8.byte_length() };
+        // Convert any '\r\n', \r, <PS> or <LS> to \n
+        bool previous_was_cr = false;
+        for (u32 code_point : source_utf8) {
+            if (previous_was_cr && code_point != '\n')
+                builder.append('\n');
+            previous_was_cr = false;
+
+            switch (code_point) {
+            case '\r':
+                previous_was_cr = true;
+                break;
+            case 0x2028:
+            case 0x2029:
```

Comment:
We can use the existing constants from Libraries/LibJS/Token.h here: `LINE_SEPARATOR` and `PARAGRAPH_SEPARATOR`. That would also make it obvious this code points are a JS requirement.